Relative API Calls (#62)

NoopDog opened the pull request:
@coverbeck I left the ability to specify the Azul Facet service via the BW_DATA_URL environment variable intact.  However, if this is not set, it now assumes a relative url to request the facets and other config.

To support running like this locally, I modified the Angular proxy and added a 2nd HTTP-Proxy. Basically any relative calls to /api/v1 will be forwarded by angular to port 3001. The HTTP-Proxy server running on 3001 will then forward to `https://ucsc-cgp.org`. The url can be changed in proxy.js

Let me know if you want to discuss. 

Cheers,
D

NoopDog committed:
Relative API Calls (#62)

coverbeck commented on README.md:
````diff
@@ -57,17 +50,26 @@ To start Express, run the following from the root directory:
+To support this, on localhost, run the proxy.json file with `node proxy.json` This will start a proxy server at port 3001. ```proxy.conf.json``` in ```/spa``` is configured to forward requests to /api to the local node.js server at port 3000 and to send /api/v1 to requests to the proxy listening at port 3001.
 
-To run the tests through the browser with a file watcher:
+The HTTP-Proxy on port 3001 is configured to send all requests to https://ucsc-cgp.org but this can be chaned as required.
````
Typo: chan**g**ed

coverbeck replied:
Also, change to `https://commons.ucsc-cgp-dev.org`.

coverbeck commented on README.md:
````diff
@@ -57,17 +50,26 @@ To start Express, run the following from the root directory:
+If this value is not set, the Boardwalk Angular client will assume the Facet Server is at the same url that served the front end, and make a relative request to /api/v1 to retrieve the facets and other configuration. 
 
-    npm test:headless
+To support this, on localhost, run the proxy.json file with `node proxy.json` This will start a proxy server at port 3001. ```proxy.conf.json``` in ```/spa``` is configured to forward requests to /api to the local node.js server at port 3000 and to send /api/v1 to requests to the proxy listening at port 3001.
````
1. That should be `node proxy.js`, not "...proxy.js**on**", shouldn't it?
2. Could you instead add a script in package.json to do this? I haven't run the UI in a while, and so simple things like running node, even though it seems pretty obvious how to do it (although maybe the next step is my fault, so maybe not), that it would just be nice to say `npm run proxy` or something like that. Just a suggestion; if you prefer it this way, then that's fine.
3. When I try running `node proxy.js`, I get the error below. Does http-proxy need to be added as a dev dependency to package.json?

```
node proxy.js
module.js:478
    throw err;
    ^

Error: Cannot find module 'http-proxy'
    at Function.Module._resolveFilename (module.js:476:15)
    at Function.Module._load (module.js:424:25)
    at Module.require (module.js:504:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/Users/charlesoverbeck/git/DataBiosphere/cgp-boardwalk/proxy.js:9:17)
    at Module._compile (module.js:577:32)
```

coverbeck commented on README.md:
```diff
@@ -57,17 +50,26 @@ To start Express, run the following from the root directory:
+// Create a HTTP Proxy server with a HTTPS target
+//
+httpProxy.createProxyServer({
+    target: 'https://ucsc-cgp.org',
```
Given that this is the feature/commons branch, I think you should be default point to our Commons deploy instance, `https://commons.ucsc-cgp-dev.org`. This will eventually matter, as only the Commons instance will support the new endpoints, e.g., `/me`, that you will be relying on.

coverbeck commented on README.md:
```diff
@@ -57,17 +50,26 @@ To start Express, run the following from the root directory:
+    target: 'https://ucsc-cgp.org',
+    agent  : https.globalAgent,
+    headers: {
+        host: 'ucsc-cgp.org'
```
In line with previous comment, I think this should default to `commons.ucsc-cgp-dev.org`

NoopDog replied:
Ok I was not aware of this updated endpoint will adjust.

coverbeck commented on spa/proxy.conf.json:
```diff
@@ -1,6 +1,13 @@
   "/api": {
     "target": "http://localhost:3000",
     "secure": false
+
```
Just checking, because I don't have the full picture for this in my head -- should `/api/v1` and `/api` go to different ports?

NoopDog replied:
Yes they ports need to be different. The Node.js back end (which will go  away is on 3000).  Facet Service Proxy is on 3001.

coverbeck replied:
👍

coverbeck commented on README.md:
````diff
@@ -57,17 +50,26 @@ To start Express, run the following from the root directory:
+
 
-Run the following from the `spa` directory:
+The URL for the back end Azul Facet Service is configurable from the environment property ```BW_DATA_URL```. This can be set to the url of the server running the back end in the Gruntfile to run locally or on the boardwalk node.js server.
````
I would prefer `Azul Webservice` over `Azul Facet Service`, but I don't feel too strongly about it.

coverbeck commented:
@theathorn I had reviewed the PR and reassigned to @NoopDog to respond to my review.

Thanks,

Charles

NoopDog commented:
OK Thanks @coverbeck will update and resubmit. Thx

NoopDog committed:
pr updates

NoopDog committed:
update readme

NoopDog commented:
Hi @coverbeck I believe I  have addressed the updates. Let me know if you have any questions. I have:

1. Added http-proxy to dev dependancies
1. updated the default back end to the commons url provided.
1. updated the readme for the commons url 
1. Attempted to address the remaining issues in the comments above

Cheers, 
D

NoopDog committed:
start proxy on npm start

coverbeck reviewed (approved):
LGTM, although you still need to run the node.js server. But I guess you have to until we give you a backend with the new endpoints...

I'm assuming you'll merge it; let me know if you're instead expecting me to merge it, so it doesn't slip through the cracks. :)

Thanks,

Charles

theathorn commented:
@coverbeck wrote: LGTM, although you still need to run the node.js server. But I guess you have to until we give you a backend with the new endpoints...

I'm assuming you'll merge it; let me know if you're instead expecting me to merge it, so it doesn't slip through the cracks.
